docs/testing.md

```diff
@@ -0,0 +1,130 @@
+## Purpose
+
+This document describes our approach to testing for the 10101 project.
+
+## Assumptions
+
+Here we list a non-exhaustive set of assumptions that influenced our findings:
+
+### Common Rust code between `app` and `coordinator`
+
+There will be a number of crates that will be used as dependencies in both `app` and `coordinator`.
+
+### No initial effort to keep different versions of `app` and `coordinator` compatible
+
+The `coordinator` will not be backwards-compatible with older versions of the `app` until later in the development of 10101.
+This means that we will not devote effort to test backwards compatibility between these two systems.
+
+### The `coordinator` is treated as a single, multi-purpose system
+
+In the past, we have modularised `coordinator`-like systems into several components for architectural and strategic reasons.
+That may very well still be the case for 10101, but at this stage we lack the insight to split this system.
+Once that design has been completed we can revise this document accordingly.
+
+### Common `10101-lightning-node` dependency between `app` and `coordinator`
+
+This `10101-lightning-node` is one of the aforementioned Rust crates shared between these two systems.
+It represents the custom Lightning Node built using (a fork of) `rust-lightning` and `rust-dlc`.
+
+### The `app`'s state will be controlled by its Rust backend
+
+Thanks to this we should be able to write system tests against the `app`'s API without involving the frontend.
+
+### The architecture will be similar to this component diagram
+
+This diagram does include some of the assumptions listed above:
+
+![Assumed component diagram](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/get10101/10101/d7619613324df3ce5c0f0a35d97bf24323f9515b/docs/diagrams/testing-component-diagram.puml)
+
+## Strategy
+
+In this section we describe the strategy that we recommend to follow when writing tests for this project.
+Each subsection corresponds to a section of the codebase or a combination thereof.
+
+### Common crates, `app` backend and `coordinator`
+
+All this code is written in Rust.
+
+#### Unit tests
+
+Good unit testing patterns apply.
+In particular, we strive for these tests to be small, structured[^1] and fast.
+Unit tests will ideally only be written for pure functions.
+
+They can be handy as API snapshot tests. For example, to ensure that a `serde` implementation doesn't change.
+
+#### Integration tests
+
+These are tests that depend on multiple components (perhaps across module boundaries).
+Additionally, this kind of tests usually involve mocked state.
+
+We are aiming to not spend too much time maintaining Rust integration tests, in favour of more comprehensive tests.
```

Comment:
The way how I interpret this: 
- you prefer writing system tests over integration tests (with mocked data).

Is this not the path we were taking in the past where we had an extensive test suit of long and slowly running tests? 

I would be in favor of having maximum a handful of system tests which need a test environment to be spun up (e.g. docker-compose). For example, these could cover only the main paths (aka happy path). Any edge case scenario I would cover on a lower level with mocked data for fast test execution.

Reply:
We discussed this topic briefly offline. These are my takeaways from the meeting:

- Here we are defining a testing strategy, but we should give ourselves an opportunity to revise it over time. With this in mind, I take responsibility to schedule a "Testing strategy retro" in about a month's time.
- As context, we agree that our ItchySats actor tests were not good enough: flaky, expensive to maintain and with long execution times.
- We agree that we must not impose long test execution times on devs or CI.
- We agree that we should consider not running all tests of all kinds on every push to the repository. Although this concern can be addressed independently in #20, it can guide us in choosing how we should write the test themselves.
- We want to explore the long-discussed possibility of running system tests that start from an in-between state, in the interest of shortening text execution times. This could severely help in allowing us to test edge case scenarios via system tests, perhaps addressing Phillip's concern mentioned in the previous comment.
- We agree that this document does not represent a clear plan of action in its current state. @da-kami volunteered to create a document that more clearly and succinctly specifies our current testing strategy.

As a side note, Daniel expressed that the way we model things has a very big effect on the quality and usability of our unit tests. He used the `Cfd` struct as an example of a bad unit, which was very hard to unit test.